smqtk_detection/impls/detect_image_objects/resnet_frcnn.py

```diff
@@ -17,61 +19,77 @@
-        :param img_batch_size: Batch size in images for inferences.
-        :param use_cuda: Attempt to use a cuda device for inferences. If no
-            device is found, CPU is used.
-        """
-
         self.box_thresh = box_thresh
         self.num_dets = num_dets
         self.img_batch_size = img_batch_size
+        self.use_cuda = use_cuda
 
-        self.model = models.detection.fasterrcnn_resnet50_fpn(
-            pretrained=True,
-            progress=False,
-            box_detections_per_img=num_dets,
-            box_score_thresh=box_thresh
-        )
-
-        self.model = self.model.eval()
-
-        if torch.cuda.is_available() and use_cuda:
-            self.use_cuda = True
-            self.model = self.model.cuda()
-        else:
-            self.use_cuda = False
+        # Set to None for lazy loading later.
+        self.model: torch.nn.Module = None  # type: ignore
 
+        # The model already has normalization and resizing baked into the
+        # layers.
         self.model_loader = transforms.Compose([
-            transforms.ToPILImage(),
+            # transforms.ToPILImage(),
             transforms.ToTensor(),
         ])
 
-        self.model.roi_heads.postprocess_detections = MethodType(_postprocess_detections, self.model.roi_heads)
+    def get_model(self) -> "torch.nn.Module":
+        """Lazy load the torch model in an idempotent manner."""
+        model = self.model
+        if model is None:
+            model = models.detection.fasterrcnn_resnet50_fpn(
+                pretrained=True,
+                progress=False,
+                box_detections_per_img=self.num_dets,
+                box_score_thresh=self.box_thresh
+            )
+            model = model.eval()
+            if self.use_cuda:
+                if torch.cuda.is_available():
+                    model = model.cuda()
+                else:
+                    LOG.warning("Use of CUDA requested, but not available. "
+                                "Proceeding with model NOT transitioned to "
+                                "CUDA.")
+            model.roi_heads.postprocess_detections = (
+                MethodType(_postprocess_detections, model.roi_heads)
+            )
+            # store the loaded model for later return.
+            self.model = model
+        return model
 
     def detect_objects(
         self,
         img_iter: Iterable[np.ndarray]
     ) -> Iterable[Iterable[Tuple[AxisAlignedBoundingBox, Dict[Hashable, float]]]]:
+        formatted_dets = []  # AxisAlignedBoundingBox detections to return
+        model = self.get_model()
 
-        formatted_dets = []       # AxisAlignedBoundingBox detections to return
-
-        img_tensors = [self.model_loader(img) for img in img_iter]
+        img_list = list(img_iter)
+        img_tensors = [self.model_loader(img) for img in img_list]
 
         if self.use_cuda:
             img_tensors = [tensor.cuda() for tensor in img_tensors]
```

Comment:
This will cause an error if `use_cuda` is true but there is no cuda device available because the model weights and input will be on different devices. Perhaps you could set `use_cuda` to false in `get_model()` when no device is found.

Reply:
I did not re-assign the `self.cuda` parameter thinking that it would be better to not perturb what was input so that the output `get_config` is reflective of what was input, while the logic would be idempotent (modulo different environments) in its interpretation of those parameters. Of course, now that I think about it, it might just be better to straight up error if you set use-cuda and there is no device to use...

A second error that you just highlighted is that we should not be using `.cuda` and instead use `to(device=...)`.